transformer_engine/pytorch/csrc/extensions/softmax.cu

```diff
@@ -245,3 +247,357 @@ at::Tensor scaled_aligned_causal_masked_softmax_backward(at::Tensor output_grad_
 
   return output_grads;
 }
+
+/***************************************************************************************************
+ * Support memory efficient cross entropy for Megatron-LM
+ **************************************************************************************************/
+template <typename dtype, int BlockSize>
+void __global__ CrossEntropyFwdSumExpKernel(float* sum_exp_logits_ptr,
+                                            dtype* vocab_parallel_logits_ptr, float* logits_max_ptr,
+                                            size_t n_dim) {
+  /***
+    1024 | 1
+    7 | 1016 | 2
+    6 | 1016 | 3
+
+    For example:
+    1024 | 1 -> [0,1023] [1024]
+    7 | 1016 | 2 -> [0, 6], [7,1022], [1023,1024]
+    ***/
+
+  /***
+    Thread model: size_t grid = rows;
+    One block is responsible for one row.
+    ***/
+  size_t rowIdx = blockIdx.x;
+  size_t tid = threadIdx.x;
+  if (tid >= n_dim) return;
+
+  size_t cur_vocab_parallel_logits_ptr_begin = rowIdx * n_dim;  // 0, 1025
+  size_t cur_vocab_parallel_logits_ptr_end =
+      rowIdx * n_dim + n_dim;  //cur_vocab_parallel_logits_ptr_end = 1025, 2050
+
+  size_t end_mol_num = cur_vocab_parallel_logits_ptr_end % 8;  //end_mol_num = 1, end_mol_num = 2
+  size_t begin_mol_num = n_dim - end_mol_num;  //begin_mol_num = 1024, begin_mol_num = 1023
+
+  //valid range for evry row is [begin_offset, end_offset]
+  size_t begin_offset = begin_mol_num % 8;      //begin_offset = 0, begin_offset = 7
+  size_t end_offset = n_dim - end_mol_num - 1;  //end_offset = 1023, end_offset = 1022
+
+  float cur_row_max = logits_max_ptr[rowIdx];
+  float cur_thread_exp_sum = 0.0;
+  float row_item = 0.0;
+
+  typedef cub::BlockReduce<float, BlockSize> BlockReduceT;
+  __shared__ typename BlockReduceT::TempStorage temp_storage;
+
+#pragma unroll
+  for (size_t i = begin_offset + tid * 8; i <= end_offset - 7; i += 8 * BlockSize) {
+    {
+      int4 int4_arr = *reinterpret_cast<int4*>(
+          &vocab_parallel_logits_ptr[cur_vocab_parallel_logits_ptr_begin + i]);
+      dtype* bf_16_p = reinterpret_cast<dtype*>(&int4_arr);
+#pragma unroll
+      for (int k = 0; k < 8; k++) {
```

Comment:
You assume here that dtype is 2B long - could we generalize this, e.g. with 
```suggestion
      for (int k = 0; k < sizeof(int4)/sizeof(dtype); k++) {
```

Reply:
Hi ptrendx, in the logic of the corresponding Megatron-LLM, its input can only be bfloat16, and we use int4 to vectorize the read, so we can assume that the loop condition here is 8

Reply:
Well, sure, but Transformer Engine is used not only with Megatron-LM, so we should not make unnecessary assumptions about the usage (especially since it does not actually cost us anything to be more general as the logic change is very simple). Also, such hardcoding could the lead to hard-to-track errors if something actually changes in the datatypes used.